components/num-util/src/fixed_decimal.rs

```diff
@@ -0,0 +1,517 @@
+use smallvec::SmallVec;
+
+use std::convert::TryFrom;
+use std::ops::RangeInclusive;
+
+use std::cmp;
+use std::string::ToString;
+
+use num_integer::Integer;
+use num_traits::{FromPrimitive, ToPrimitive, Unsigned};
+use static_assertions::const_assert;
+
+use super::uint_iterator::UintIterator;
+
+#[cfg(test)]
+use num_bigint::BigUint;
+#[cfg(test)]
+use std::convert::TryInto;
+
+// FixedDecimal assumes usize (digits.len()) is at least as big as a u16
+const_assert!(std::mem::size_of::<usize>() >= std::mem::size_of::<u16>());
+
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    /// The magnitude or number of digits exceeds the limit of the FixedDecimal. The highest
+    /// magnitude of the most significant digit is std::i16::MAX, and the lowest magnitude of the
+    /// least significant digit is std::i16::MIN.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use icu_num_util::FixedDecimal;
+    /// use icu_num_util::fixed_decimal::Error;
+    ///
+    /// let mut dec1 = FixedDecimal::from(123);
+    /// assert_eq!(Error::Limit, dec1.adjust_magnitude(std::i16::MAX).unwrap_err());
+    /// ```
+    Limit,
+}
+
+/// A struct containing decimal digits with efficient iteration and manipulation by magnitude
+/// (power of 10). Supports a mantissa of non-zero digits and a number of leading and trailing
+/// zeros, used for formatting and plural selection.
+///
+/// You can create a FixedDecimal from an unsigned integer or an isize.
+///
+/// # Examples
+///
+/// ## From an isize
+///
+/// ```
+/// use icu_num_util::FixedDecimal;
+///
+/// let dec1 = FixedDecimal::from(250);
+/// assert_eq!("250", dec1.to_string());
+/// ```
+///
+/// ## From a BigUint
+///
+/// ```
+/// use icu_num_util::FixedDecimal;
+/// use icu_num_util::fixed_decimal::UnsignedWrapper;
+/// use num_bigint::BigUint;
+/// use std::convert::TryFrom;
+///
+/// let input = BigUint::parse_bytes(b"98765432109876543210", 10).unwrap();
+/// let dec2 = FixedDecimal::try_from(UnsignedWrapper(input)).unwrap();
+/// assert_eq!("98765432109876543210", dec2.to_string());
+/// ```
+#[derive(Debug, Clone, PartialEq)]
+pub struct FixedDecimal {
+    /// List of digits; digits[0] is the most significant.
+    ///
+    /// Invariants:
+    /// - Must not include leading or trailing zeros
+    /// - Length must not exceed (magnitude - lower_magnitude + 1)
+    // TODO: Consider using a nibble array
+    digits: SmallVec<[u8; 8]>,
```

Comment:
It seems "obvious" to me that `SmallVec` is the right tool for the job compared to `Vec` and `Box<[u8]>`.  Do you have experience that suggests that `SmallVec` *isn't* always the objectively best option for vectors with a small number of bytes on the stack?  I personally haven't seen or read anything that has given me pause when picking SmallVec for a short array of bytes.

To answer your questions specifically:

1. I have previously compared SmallVec to Vec (actually SmallStr to String), and I found negligible difference in code size, with SmallStr coming out slightly smaller than String.  I haven't looked at speed or memory, only code size, and I haven't done it for this specific use case.  For optimal velocity, I would rather compare things like SmallVec vs Vec in a vacuum and make general recommendations.
2. Mutable operations will likely be very common on this class, specifically when we implement rounding behavior.

Reply:
> It seems "obvious" to me that SmallVec is the right tool for the job compared to Vec and Box<[u8]>. Do you have experience that suggests that SmallVec isn't always the objectively best option for vectors with a small number of bytes on the stack?

Yes. In several scenarios `SmallVec` was slower than `Vec`.
Now, I understand that `SmallVec` gives you stack, but I'd test it yourself.

> Mutable operations will likely be very common on this class, specifically when we implement rounding behavior.

I hear you, so mutability will be common. Can you give me example of use case where one would allocate a `FixedDecimal` and then manipulate it?
I'm ok with imaginary based on your experience. Would it be more of a hot-code in other ICU methods, or userland code that would use it?

Reply:
1. OK, since I have a benchmark file now, it should be easy to do a quick speed check.  I'll get to that on Monday or Tuesday.
2. I anticipate that users will generally create a FixedDecimal and give it (by value) to NumberFormat, which will then perform mutative operations like rounding to the nearest cent or scaling and rounding according to compact notation preferences.  I do not anticipate that userland code would generally mutate a FixedDecimal.

I went back and forth multiple times on whether to traitify FixedDecimal rather than giving a concrete struct.  I acknowledge that a trait seems to make more sense, but I settled on a struct.  I did not spend the time to document why I did this.  I can rehash that decision if you dispute it.

Reply:
Not disputing, I'm all in favor of starting with a concrete type even for things we may later add a trait for.

I'm trying to understand the use.
From what you're saying the typical use will be to use `From` from some source value to a `FixedDecimal` that then will be passed to some `NumberFormat` related code that will adjust the rounding and adjusting the visible digits.

That convinces me that mutability is a hot code for this, and that `Box<[u8]>` is not the right use case.

Reply:
I did some initial benchmarks and `SmallVec` does seem to be the most efficient!